src/06-mocking-node-api/index.test.ts

```diff
@@ -1,52 +1,114 @@
-// Uncomment the code below and write your tests
-// import { readFileAsynchronously, doStuffByTimeout, doStuffByInterval } from '.';
+import { readFileAsynchronously, doStuffByTimeout, doStuffByInterval } from '.';
+import fs from 'fs';
+import fsPromises from 'fs/promises';
+import path from 'path';
 
 describe('doStuffByTimeout', () => {
+  let timeoutSpy: jest.SpyInstance;
+  const callback = jest.fn();
+  const timeout = 10;
+
   beforeAll(() => {
     jest.useFakeTimers();
   });
 
+  beforeEach(() => {
+    timeoutSpy = jest.spyOn(global, 'setTimeout');
+  });
+
+  afterEach(() => {
+    jest.clearAllTimers();
+  });
+
   afterAll(() => {
     jest.useRealTimers();
   });
 
   test('should set timeout with provided callback and timeout', () => {
-    // Write your test here
+    doStuffByTimeout(callback, timeout);
+
+    expect(timeoutSpy).toBeCalledWith(callback, timeout);
   });
 
   test('should call callback only after timeout', () => {
-    // Write your test here
+    doStuffByTimeout(callback, timeout);
+    jest.runAllTimers();
+
+    expect(callback).toBeCalled();
   });
 });
 
 describe('doStuffByInterval', () => {
+  let intervalSpy: jest.SpyInstance;
+  const callback = jest.fn();
+  const interval = 10;
+
   beforeAll(() => {
     jest.useFakeTimers();
   });
 
+  afterEach(() => {
+    jest.clearAllTimers();
+  });
+
+  beforeEach(() => {
+    intervalSpy = jest.spyOn(global, 'setInterval');
+  });
+
   afterAll(() => {
     jest.useRealTimers();
   });
 
   test('should set interval with provided callback and timeout', () => {
-    // Write your test here
+    doStuffByInterval(callback, interval);
+
+    expect(intervalSpy).toBeCalledWith(callback, interval);
   });
 
   test('should call callback multiple times after multiple intervals', () => {
-    // Write your test here
+    const count = 5;
+
+    doStuffByInterval(callback, interval);
+    jest.advanceTimersToNextTimer(count);
+
+    expect(callback).toBeCalledTimes(count);
   });
 });
 
 describe('readFileAsynchronously', () => {
+  let existSyncSpy: jest.SpyInstance;
+  let readFileSpy: jest.SpyInstance;
+  let joinSpy: jest.SpyInstance;
+
+  beforeEach(() => {
+    existSyncSpy = jest.spyOn(fs, 'existsSync');
+    readFileSpy = jest.spyOn(fsPromises, 'readFile');
+    joinSpy = jest.spyOn(path, 'join');
+  });
+
   test('should call join with pathToFile', async () => {
-    // Write your test here
+    const pathToFile = 'path';
+
+    await readFileAsynchronously(pathToFile);
+
+    expect(joinSpy.mock.calls[0]).toContain(pathToFile);
```

Comment:
Assignment tell "Remember that the tests should not interact with the actual file system and should not rely on real-time!"

only 3 point for this test because you use real nodejs function

if you put `console.log('nodejs "join" function work proof: ', join('111', '222'));` in readFileAsynchronously

you will see
```
  console.log
    nodejs "join" function work proof:  111\222
```

it is a mistake

you sould use
`joinSpy.getMockImplementation(() => ...)`
or better
`jest.mock('path');`